### ISSUE TITLE
CJK word boundaries

### ISSUE THREAD

**Issue by justinmc** [flutter/flutter#123065]
Moving the cursor by word does not produce correct results when the text contains Chinese characters.  Currently, it seems to move character-by-character, but instead it should move word by word.

| Incorrect Flutter behavior | Correct Mac behavior |
| --- | --- |
| <video src="https://user-images.githubusercontent.com/389558/213824760-b7b35d2b-f762-4dfa-92c3-c1d6de71617e.webm)" />| <video src="https://user-images.githubusercontent.com/389558/226982482-9a99d1ba-ddf4-420b-adaf-1af56d906b2a.mov" /> |

### Steps to reproduce

  1. Run an app with a text input, such as the one below.
  2. Enter some Chinese text containing words that are more than 1 character in length, such as the example text below.
  3. Press control+arrow left/right (or option+arrow left/right on Mac).

Expected results: The cursor moves left/right to the nearest word boundary.
Actual results: The cursor moves by one character.

<details>

<summary>Example app</summary>

```dart
import 'package:flutter/material.dart';

void main() {
  runApp(const MyApp());
}

class MyApp extends StatelessWidget {
  const MyApp({super.key});

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'Flutter Demo',
      home: Scaffold(
        appBar: AppBar(
          title: const Text("Demo"),
        ),
        body: const Center(
          child: Column(
            mainAxisAlignment: MainAxisAlignment.center,
            children: <Widget>[
              TextField(
              ),
            ],
          ),
        ),
      ),
    );
  }
}
```

</details>

<details>

<summary>Example Chinese text</summary>

1894年，孫文成立興中會。1895年，興中會發動乙未廣州起義。1905年，中國同盟會在日本东京組成，孫被推為总理；確定「驅除韃虜，恢復中华，建立民國，平均地權」的革命政綱，提出三民主義學說；創辦《民報》，宣傳革命，同當時的中國改良派激烈論戰。

Source: https://github.com/flutter/flutter/issues/106472#issuecomment-1465077065

</details>

### Next steps

Does [icu](https://github.com/unicode-org/icu) have built-in support for finding Chinese word breaks?  Does SkParagraph?

Flutter may need to include a Chinese dictionary at some level for this to work.  This must be able to be tree-shaken out, because I assume the size would be non-trivial.

If that is not practical, maybe we should move between punctuation marks instead?

### Related

This is a continuation of https://github.com/flutter/flutter/issues/106472.

**Comment by gnprice** [comment]
Similar remarks apply to Japanese. I think probably any solution we find for Chinese will come with a solution for Japanese and vice versa, so there isn't much separate work to do for one vs. the other; mentioning it here mainly so the issue matches the term "japanese" in searches.

Here's some example Japanese text, again from [a Wikipedia article](https://ja.wikipedia.org/wiki/語):

> たとえば日本語の辞典で、「考えた」「考えない」「考えれば」「考えろ」などとは記載されず、「考える」だけが記載される。

I just tried the repro steps and confirmed they give the same results on that text as for Chinese.

**Comment by tgucio** [comment]
Typically Japanese, Chinese, Thai, Khmer, Lao and Myanmar/Burmese would require dictionary-based word breaking algo and indeed the data included in ICU is rather large.
https://unicode-org.github.io/icu/userguide/boundaryanalysis/

I guess a simplistic approach to have something to start with but w/o dict, could involve:
1) stop at all spaces, including ZWSP
2) stop at all punctuation marks
3) stop at numbers
Maybe to cover all 1-3 stopping at all characters whose script Unicode property is different from the current one could be a good start.

FWIW, Flutter Web engine has a good word breaker implementing Unicode rules (naturally w/o dict): https://github.com/flutter/engine/blob/main/lib/web_ui/lib/src/engine/text/word_breaker.dart

**Comment by tgucio** [comment]
BTW, just for Lao there's this old project:
https://code.google.com/archive/p/lao-dictionary/source/default/source

**Comment by gnprice** [comment]
> I guess a simplistic approach to have something to start with but w/o dict, could involve:

This sounds similar to the gedit behavior described at https://github.com/flutter/flutter/issues/106472#issuecomment-1399082708 .

To copy into this thread what I said there: that behavior isn't great (it means moving past entire long clauses)… but it's still probably worthwhile to make Ctrl+Left/Right do that, instead of one character at a time, if we think we won't be able to do real word boundaries anytime soon. After all, if you want a character at a time you can always use plain left/right without Ctrl.


> Maybe to cover all 1-3 stopping at all characters whose script Unicode property is different from the current one could be a good start.

Hmm, will that do it? I'd have thought that punctuation like 。 and 、 that are used in Japanese, for example, would be in the same script as Japanese text.

**Comment by gnprice** [comment]
> BTW (to get an idea of sizes), just for Lao there's this old project:
> https://code.google.com/archive/p/lao-dictionary/source/default/source

Cool. Just to put the numbers explicitly in the thread: that Lao dictionary is
205kB uncompressed
46kB compressed in that zip archive
36kB compressed with `xz`

**Comment by tgucio** [comment]
> Hmm, will that do it? I'd have thought that punctuation like 。 and 、 that are used in Japanese, for example, would be in the same script as Japanese text.
Good question there. Just checked in Unicode data (https://unicode.org/Public/UNIDATA/UnicodeData.txt and https://unicode.org/Public/UNIDATA/Scripts.txt):

- "。" is `3002;IDEOGRAPHIC FULL STOP;Po;0;ON;;;;;N;IDEOGRAPHIC PERIOD;;;;`
- it's assigned the "Common" script property: `3001..3003    ; Common # Po   [3] IDEOGRAPHIC COMMA..DITTO MARK`

**Comment by tgucio** [comment]
Just ran a little test with the word_breaker.dart stolen from the Engine web_ui - here's what the standard Unicode word breaking w/o dicts yields:

```
Wszyscy ludzie rodzą się wolni i równi pod względem swej godności i swych praw. => |Wszyscy| |ludzie| |rodzą| |się| |wolni| |i| |równi| |pod| |względem| |swej| |godności| |i| |swych| |praw|.|
Όλοι οι άνθρωποι γεννιούνται ελεύθεροι και ίσοι στην αξιοπρέπεια και τα δικαιώματα. => |Όλοι| |οι| |άνθρωποι| |γεννιούνται| |ελεύθεροι| |και| |ίσοι| |στην| |αξιοπρέπεια| |και| |τα| |δικαιώματα|.|
Всі люди народжуються вільними і рівними у своїй гідності та правах. => |Всі| |люди| |народжуються| |вільними| |і| |рівними| |у| |своїй| |гідності| |та| |правах|.|
כל בני אדם נולדו בני חורין ושווים בערכם ובזכויותיהם. => |כל| |בני| |אדם| |נולדו| |בני| |חורין| |ושווים| |בערכם| |ובזכויותיהם|.|
يولد جميع الناس أحرارًا متساوين في الكرامة والحقوق. => |يولد| |جميع| |الناس| |أحرارًا| |متساوين| |في| |الكرامة| |والحقوق|.|
सभी मनुष्यों को गौरव और अधिकारों के मामले में जन्मजात स्वतन्त्रता और समानता प्राप्त है । => |सभी| |मनुष्यों| |को| |गौरव| |और| |अधिकारों| |के| |मामले| |में| |जन्मजात| |स्वतन्त्रता| |और| |समानता| |प्राप्त| |है| |।|
มนุษย์ทั้งหลายเกิดมามีอิสระและเสมอภาคกันในเกียรติศักด[เกียรติศักดิ์]และสิทธิ => |ม|นุ|ษ|ย์|ทั้|ง|ห|ล|า|ย|เ|กิ|ด|ม|า|มี|อิ|ส|ร|ะ|แ|ล|ะ|เ|ส|ม|อ|ภ|า|ค|กั|น|ใ|น|เ|กี|ย|ร|ติ|ศั|ก|ด|[|เ|กี|ย|ร|ติ|ศั|ก|ดิ์|]|แ|ล|ะ|สิ|ท|ธิ|
ມະນຸດເກີດມາມີສິດເສລີພາບ ແລະ ສະເໝີໜ້າກັນໃນທາງກຽດຕິສັກ ແລະ ທາງສິດ. => |ມ|ະ|ນຸ|ດ|ເ|ກີ|ດ|ມ|າ|ມີ|ສິ|ດ|ເ|ສ|ລີ|ພ|າ|ບ| |ແ|ລ|ະ| |ສ|ະ|ເ|ໝີ|ໜ້|າ|ກັ|ນ|ໃ|ນ|ທ|າ|ງ|ກ|ຽ|ດ|ຕິ|ສັ|ກ| |ແ|ລ|ະ| |ທ|າ|ງ|ສິ|ດ|.|
មនុស្សទាំងអស់ កើតមកមានសេរីភាព និងសមភាព ក្នុងផ្នែកសេចក្ដីថ្លៃថ្នូរនិងសិទ្ធិ។ => |ម|នុ|ស្|ស|ទាំ|ង|អ|ស់| |កើ|ត|ម|ក|មា|ន|សេ|រី|ភា|ព| |និ|ង|ស|ម|ភា|ព| |ក្|នុ|ង|ផ្|នែ|ក|សេ|ច|ក្|ដី|ថ្|លៃ|ថ្|នូ|រ|និ|ង|សិ|ទ្|ធិ|។|
人人生而自由,在尊严和权利上一律平等。他们赋有理性和良心,并应以兄弟关系的精神相对待。 => |人|人|生|而|自|由|,|在|尊|严|和|权|利|上|一|律|平|等|。|他|们|赋|有|理|性|和|良|心|,|并|应|以|兄|弟|关|系|的|精|神|相|对|待|。|
すべての人間は、生まれながらにして自由であり、かつ、尊厳と権利とについて平等である。 => |す|べ|て|の|人|間|は|、|生|ま|れ|な|が|ら|に|し|て|自|由|で|あ|り|、|か|つ|、|尊|厳|と|権|利|と|に|つ|い|て|平|等|で|あ|る|。|
```

So perhaps indeed this + script property breaking for languages that would require dicts could be a starting point.

<details>
  <summary>Test code</summary>

```
    const List<String> texts = <String>[
      'Wszyscy ludzie rodzą się wolni i równi pod względem swej godności i swych praw.',
      'Όλοι οι άνθρωποι γεννιούνται ελεύθεροι και ίσοι στην αξιοπρέπεια και τα δικαιώματα.',
      'Всі люди народжуються вільними і рівними у своїй гідності та правах.',
      'כל בני אדם נולדו בני חורין ושווים בערכם ובזכויותיהם.',
      'يولد جميع الناس أحرارًا متساوين في الكرامة والحقوق.',
      'सभी मनुष्यों को गौरव और अधिकारों के मामले में जन्मजात स्वतन्त्रता और समानता प्राप्त है ।',
      'มนุษย์ทั้งหลายเกิดมามีอิสระและเสมอภาคกันในเกียรติศักด[เกียรติศักดิ์]และสิทธิ',
      'ມະນຸດເກີດມາມີສິດເສລີພາບ ແລະ ສະເໝີໜ້າກັນໃນທາງກຽດຕິສັກ ແລະ ທາງສິດ.',
      'មនុស្សទាំងអស់ កើតមកមានសេរីភាព និងសមភាព ក្នុងផ្នែកសេចក្ដីថ្លៃថ្នូរនិងសិទ្ធិ។',
      '人人生而自由,在尊严和权利上一律平等。他们赋有理性和良心,并应以兄弟关系的精神相对待。',
      'すべての人間は、生まれながらにして自由であり、かつ、尊厳と権利とについて平等である。',
    ];

    for (String text in texts) {
      int index = 0, prevIndex = 0;
      final StringBuffer sb = StringBuffer('|');
      do {
        index = WordBreaker.nextBreakIndex(text, index);
        sb.write('${text.substring(prevIndex, index)}|');
        prevIndex = index;
      } while (index < text.length);
      print('$text => $sb');
    }
```
</details>

**Comment by gnprice** [comment]
Looks like that does a good job on the languages that use spaces and don't require dicts. Is improvement needed there? At a quick test, it seems like Ctrl+Left/Right are already doing fine for those, modulo #78660. (I tried the examples in Greek, Devanagari, and Arabic script.)

**Comment by tgucio** [comment]
Well, the test code output is from the `WordBreaker` that's in the engine's web_ui which is a complete Unicode implementation (https://unicode.org/reports/tr29/#Word_Boundary_Rules). What we have in the framework (`WordBoundary`) does not implement all the rules (it doesn't use Unicode char properties). So this could be the first improvement.

The second improvement, for Chinese/Han and Japanese/Kanji, could be to use the Unicode script properties to stop on punctuation. This would be non-standard (while no dicts are available) and I guess care should be taken not to break Unicode rules such as https://unicode.org/reports/tr29/#WB11).

**Comment by gnprice** [comment]
> This would be non-standard (while no dicts are available) and I guess care should be taken not to break Unicode rules such as https://unicode.org/reports/tr29/#WB11).

Hmm, I see. Looking at that table, perhaps we can implement the "script property breaking" idea in the form of inserting additional rules into the list? It looks like that's the intended way for an implementation to tailor the behavior.

For example, looking in the CLDR tree, here's what they have [in common/segments/ja.xml](https://github.com/unicode-org/cldr/blob/8f04544c6f31812a6bfd2fecc6ca6d2fd715771a/common/segments/ja.xml) for customizing word breaks for Japanese:
```xml
		<segmentation type="WordBreak">
			<variables>
				<variable id="$Hiragana">\p{Hiragana}</variable>
				<variable id="$Hiragana">($Hiragana $FE*)</variable>
				<variable id="$Ideographic">[[\p{Ideographic}] [\u3005 \u3007 \u303B]]</variable>
				<variable id="$Ideographic">($Ideographic $FE*)</variable>
			</variables>
			<segmentRules>
				<rule id="13.3"> $Hiragana × $Hiragana </rule>
				<rule id="13.4"> $Ideographic × $Ideographic </rule>
			</segmentRules>
		</segmentation>
```

In other words, one follows all the rules in the default table, except that before you reach the rule at the very end which inserts a break if no other rule applies, you forbid breaks within runs of hiragana and runs of "ideographic", where each of those has a definition expressed in the `<variables>` section. (The default rules already forbid breaks within runs of katakana.)

If we take those rules and then add a similar one for each of Thai script, Lao script, Khmer script, and Myanmar/Burmese script, that may do it for as good as we can get short of dictionary data.

**Comment by tgucio** [comment]
Sounds good - the properties missing in UNIDATA/auxiliary/WordBreakProperty.txt seem to be coming from different data though. Ideographic would be UNIDATA/PropList.txt and Hiragana from UNIDATA/Scripts.txt I guess?

South-East asian scripts also have peculiarities;
- Thai and Lao don't use punctuation (space is used as full stop), except maybe for modern texts where you can see Latin punctuation marks: "!", "?", parentheses etc
- Khmer has periods "។" / "៕", colon "៖" and a bunch of old (rarely used?) symbols
- Burmese has diacritics similar to other Brahmic scripts: "၊", "။" plus a few others

**Comment by justinmc** [comment]
I think it would be possible to move word_breakder.dart somewhere that's accessible by both the framework and the engine (like how dart.ui is used in both).  But even that might be a non-trivial size increase for non-web apps?

CC @yjbanov or anyone else more familiar with web that might have some thoughts.

**Comment by tgucio** [comment]
The packed lookup data isn't big really. I can see it's recently been moved to https://github.com/flutter/engine/blob/main/third_party/web_unicode/lib/web_unicode/codegen/word_break_properties.dart

It's 8245 characters so I guess it'd occupy twice that in bytes.

**Comment by gspencergoog** [comment]
We've recently adopted a new version of ICU data tables for desktop, so proper handling of this will probably be more feasible now.

**Comment by Renzo-Olivares** [comment]
@gspencergoog How can we determine if those new data tables would properly break up chinese words/other languages? And would they be usable on mobile platforms?

**Comment by gspencergoog** [comment]
I actually don't know, sorry, just that the tables are now more comprehensive (I was typing the comment for folks at the Desktop Triage meeting)  On mobile the issue would be compilation size, even if it helps break them up properly.

**Comment by justinmc** [comment]
I think this is related to https://github.com/flutter/flutter/issues/59284, which is about word breaking in Korean.

**Comment by HighLiuk** [comment]
On mobile the same problem happens in the context of selectable text. When selecting a word by long press or double tap, a single (Japanese) character is selected instead.

Isn't there a workaround for this? Even if I can provide the correct word splitting (see mecab_dart, ve_dart) I can't manage to solve how to select that word between provided boundaries.
